cmd/node.go

```diff
@@ -0,0 +1,79 @@
+package cmd
+
+import (
+	"os"
+	"os/signal"
+	"syscall"
+
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/module"
+)
+
+type Node interface {
+	module.ReadyDoneAware
+
+	// Run initiates all common components (logger, database, protocol state etc.)
+	// then starts each component. It also sets up a channel to gracefully shut
+	// down each component if a SIGINT is received.
+	Run()
+
+	// ShutdownSignal returns a channel that is closed when shutdown has commenced.
+	ShutdownSignal() <-chan struct{}
+}
+
+type FlowNodeImp struct {
+	module.ReadyDoneAware
+	*NodeConfig
+	Logger zerolog.Logger
+	sig    chan os.Signal
+}
+
+func NewNode(builder NodeBuilder, cfg *NodeConfig, logger zerolog.Logger, sig chan os.Signal) Node {
+	return &FlowNodeImp{
+		ReadyDoneAware: builder,
+		NodeConfig:     cfg,
+		Logger:         logger,
+		sig:            sig,
+	}
+}
+
+// Run calls Ready() to start all the node modules and components. It also sets up a channel to gracefully shut
+// down each component if a SIGINT is received. Until a SIGINT is received, Run will block.
+// Since, Run is a blocking call it should only be used when running a node as it's own independent process.
+func (node *FlowNodeImp) Run() {
+	// initialize signal catcher
+	signal.Notify(node.sig, os.Interrupt, syscall.SIGTERM)
+
+	select {
+	case <-node.Ready():
+		node.Logger.Info().Msgf("%s node startup complete", node.BaseConfig.NodeRole)
+	case <-node.sig:
+		node.Logger.Warn().Msg("node startup aborted")
+		os.Exit(1)
+	}
```

Comment:
Same race condition here, right?

Reply:
You mean that we could have received the signal but also become ready, and we don't run the shutdown logic?

This is just a copy/paste from `scaffold.go`. Those race conditions are handled explicitly in the next PR.